crates/proposer-client/src/lib.rs

```diff
@@ -0,0 +1,109 @@
+use std::fmt::Display;
+use std::sync::Arc;
+use std::time::Duration;
+
+use serde::{Deserialize, Serialize};
+use sp1_sdk::network::proto::network::FulfillmentStatus;
+use sp1_sdk::network::B256;
+use sp1_sdk::{NetworkProver, ProverClient, SP1ProofWithPublicValues};
+
+pub use crate::config::ProposerClientConfig;
+pub use crate::error::Error;
+use crate::rpc::ProposerRpcClient;
+
+const CLUSTER_POLL_DURATION_STEP: Duration = Duration::from_secs(10);
+
+pub mod config;
+pub mod error;
+mod rpc;
+
+#[derive(Clone)]
+pub struct ProposerClient {
+    rpc: Arc<ProposerRpcClient>,
+    prover_client: Arc<NetworkProver>,
+    proving_timeout: Duration,
+}
+
+impl ProposerClient {
+    pub fn new(config: ProposerClientConfig) -> Result<Self, error::Error> {
+        let proposer_url = format!("http://{}:{}", config.proposer_host, config.proposer_port);
+        let cluster_url = format!(
+            "http://{}:{}",
+            config.sp1_cluster_host, config.sp1_cluster_port
+        );
+
+        let network_prover = ProverClient::builder()
+            .network()
+            .rpc_url(&cluster_url)
+            .build();
+
+        Ok(ProposerClient {
+            rpc: Arc::new(ProposerRpcClient::new(&proposer_url)?),
+            prover_client: Arc::new(network_prover),
+            proving_timeout: config.proving_timeout,
+        })
+    }
+
+    pub async fn request_agg_proof(&mut self, request: Request) -> Result<ProofId, Error> {
+        self.rpc
+            .request_agg_proof(request.into())
+            .await
+            .map(Into::into)
+    }
+
+    pub async fn wait_for_proof(
+        &mut self,
+        proof_id: ProofId,
+    ) -> Result<Option<SP1ProofWithPublicValues>, Error> {
+        if proof_id.0.len() != 32 {
+            return Err(Error::InvalidProofId(proof_id));
+        }
+        let request_id = B256::from_slice(proof_id.0.as_slice());
+        let mut remaining_timeout = self.proving_timeout;
```

Comment:
I see a [wait_proof](https://github.com/succinctlabs/sp1/blob/a4e023df60a43cc00ba54469966b257801732421/crates/sdk/src/network/prover.rs#L273-L311) in the api which seems to do something similar and might replace this whole function?

Reply:
Nice catch